src/gazebo_mavlink_interface.cpp

```diff
@@ -653,27 +651,45 @@ void GazeboMavlinkInterface::setMavlinkSensorOrientation(const ignition::math::V
 
 void GazeboMavlinkInterface::ImuCallback(ImuPtr& imu_message)
 {
-  std::unique_lock<std::mutex> lock(last_imu_message_mutex_);
+  std::unique_lock<std::mutex> lock(imu_received_mutex_);
 
-  const int64_t diff = imu_message->seq() - last_imu_message_.seq();
+  const int64_t diff = imu_message->seq() - last_imu_message_seq_;
   if (diff != 1 && imu_message->seq() != 0)
   {
     gzerr << "Skipped " << (diff - 1) << " IMU samples (presumably CPU usage is too high)\n";
   }
 
-  last_imu_message_ = *imu_message;
+  last_imu_message_seq_ = imu_message->seq();
+
+  ignition::math::Quaterniond q_gr = ignition::math::Quaterniond(
+    imu_message->orientation().w(),
+    imu_message->orientation().x(),
+    imu_message->orientation().y(),
+    imu_message->orientation().z());
+
+  ignition::math::Vector3d accel_b = q_FLU_to_FRD.RotateVector(ignition::math::Vector3d(
+    imu_message->linear_acceleration().x(),
+    imu_message->linear_acceleration().y(),
+    imu_message->linear_acceleration().z()));
+
+  ignition::math::Vector3d gyro_b = q_FLU_to_FRD.RotateVector(ignition::math::Vector3d(
+    imu_message->angular_velocity().x(),
+    imu_message->angular_velocity().y(),
+    imu_message->angular_velocity().z()));
+
+  SensorData::Imu imu_data;
+  imu_data.accel_b = Eigen::Vector3d(accel_b.X(), accel_b.Y(), accel_b.Z());
+  imu_data.gyro_b = Eigen::Vector3d(gyro_b.X(), gyro_b.Y(), gyro_b.Z());
+  mavlink_interface_->UpdateIMU(imu_data);
```

Comment:
Actually looking at this again, I think we need to add the id of the imu sensor here, so it can be considered when sending over in the `HIL_SENSOR` message

Reply:
You ar totally right, but i wanted to make the multi IMU structure in a separate PR. Here i really only wanted to talk about why the current imu message was handled completely different, and i restructured in such a way, that it is easy to amek it multiple IMU capable.